src/bottom-bar/use-on-complete-callback.js

```diff
@@ -0,0 +1,166 @@
+        warning: true,
+    })
+
+    return () => {
+        return Promise.all([
+            validate()
+                .then(
+                    ({
+                        commentRequiredViolations,
+                        validationRuleViolations,
+                    }) => {
+                        if (
+                            hasViolations(
+                                commentRequiredViolations,
+                                validationRuleViolations
+                            )
+                        ) {
+                            setRightHandPanel(validationResultsSidebarId)
+                        }
+                    }
+                )
+                .catch((e) => {
+                    console.error(e)
+                    // For now this can fail "silently".
+                    // ES6 Promises don't have a mechanism for reporting
+                    // multiple promise failures and we def. want to notify the
+                    // user when form completion fails as well.
+                    // It should be enough to show a warning to the user
+                    showWarningAlert(validationFailedWarningMessage)
+                }),
+            setFormCompletion({ completed: true }).catch(() => {
+                throw new Error(completingFormFailedMessage)
+            }),
+        ])
+    }
+}
+
+function useOnCompleteWithoutValidationClick() {
+    const { mutateAsync: setFormCompletion } = useSetFormCompletionMutation()
+    return () => setFormCompletion({ completed: true })
+}
+
+export default function useOnCompleteCallback() {
+    const { offline } = useConnectionStatus()
+    const { data: metadata } = useMetadata()
+    const [dataSetId] = useDataSetId()
+    const dataSet = selectors.getDataSetById(metadata, dataSetId)
+    const { validCompleteOnly } = dataSet
+    const { show: showErrorAlert } = useAlert((message) => message, {
+        critical: true,
+    })
+    const setFormCompletionMutationKey = useSetFormCompletionMutationKey()
+    const isLoading = useIsMutating(setFormCompletionMutationKey)
+    const cancelCompletionMutation = useImperativeCancelCompletionMutation()
+    const onCompleteWhenValidRequiredClick =
+        useOnCompleteWhenValidRequiredClick()
+    const onCompleteWhenValidNotRequiredClick =
+        useOnCompleteWhenValidNotRequiredClick()
+    const onCompleteWithoutValidationClick =
+        useOnCompleteWithoutValidationClick()
+
+    return () => {
+        if (isLoading) {
+            cancelCompletionMutation()
+        }
+
+        let promise
+        if (isLoading && offline) {
+            // No need to complete when the completion request
+            // hasn't been sent yet due to being offline.
+            // It's important to still perform the request online as we don't
+            // know if the mutation actually reached the server already
+            return Promise.resolve()
```

Comment:
Not sure I fully understand this. Is the mutation actually sent here? The request was just cancelled above, we should prepare a mutation in this case, no?

Reply:
No mutation is being sent. When the user is offline and we have a pending mutation, instead of adding a second pending request, we can simply cancel the pending request:

pending-complete-request -> pending-incomplete-request ---> status: incomplete
pending-complete-request -> cancel request ---> status: incomplete

Reply:
Yeah, then I think the `cancelCompletionMutation()` above should be move into this if-test.

Reply:
:ok_hand:, done